src/mqt/predictor/reward.py

```diff
@@ -69,3 +71,92 @@ def calc_qubit_index(qargs: list[Qubit], qregs: list[QuantumRegister], index: in
             return qubit_index
     error_msg = f"Global qubit index for local qubit {index} index not found."
     raise ValueError(error_msg)
+
+
+def expected_success_probability(qc: QuantumCircuit, device: Device, precision: int = 10) -> float:
+    """Calculates the expected success probability of a given quantum circuit on a given device.
+
+    Arguments:
+        qc: The quantum circuit to be compiled.
+        device: The device to be used for compilation.
+        precision: The precision of the returned value. Defaults to 10.
+
+    Returns:
+        The expected success probability of the given quantum circuit on the given device.
+    """
+    res = 1.0
+
+    # collect gate durations for operations in the circuit
+    op_times = []
+    try:
+        for instruction, qargs, _cargs in qc.data:
+            gate_type = instruction.name
+            if gate_type == "barrier" or gate_type == "id":
+                continue
+            assert len(qargs) in [1, 2]
+            first_qubit_idx = calc_qubit_index(qargs, qc.qregs, 0)
+
+            if len(qargs) == 1:  # single-qubit gate
+                if gate_type == "measure":
+                    duration = device.get_readout_duration(first_qubit_idx)
+                else:
+                    duration = device.get_single_qubit_gate_duration(gate_type, first_qubit_idx)
+                op_times.append((gate_type, [first_qubit_idx], duration, "s"))
+            else:  # multi-qubit gate
+                second_qubit_idx = calc_qubit_index(qargs, qc.qregs, 1)
+                duration = device.get_two_qubit_gate_duration(gate_type, first_qubit_idx, second_qubit_idx)
+                op_times.append((gate_type, [first_qubit_idx, second_qubit_idx], duration, "s"))
+    except ValueError:
+        logger.exception("Calculating ESP requires device with fully specified gate and readout durations.")
```

Comment:
```suggestion
        logger.exception(f"Calculating ESP requires device with fully specified gate and readout durations that is not provided for {device.name}.")
```

I suggest to add the device name here as well. Not sure whether the suggestion works, it might need some minor modification.